### PR TITLE
Refactor MemoryLayerTest

### PR DESCRIPTION
**Pull request by olpipi**
### Details:
 - Refactor MemoryLayerTest

### Tickets:
 - [CVS-111364](https://jira.devtools.intel.com/browse/CVS-111364)

### CODE REVIEW

File: src/tests/functional/shared_test_classes/include/shared_test_classes/single_op/memory.hpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#include <vector>
+
+#include "shared_test_classes/base/ov_subgraph.hpp"
+#include "ov_models/utils/ov_helpers.hpp"
+
+namespace ov {
+namespace test {
+
+using MemoryLayerTestParams = std::tuple<
+        ngraph::helpers::MemoryTransformation,   // Apply Memory transformation
```
Comment (vurusovs) [inline src/tests/functional/shared_test_classes/include/shared_test_classes/single_op/memory.hpp:19]: need to port `ngraph::helpers::MemoryTransformation`, may be done in scope of this PR

File: src/tests/functional/shared_test_classes/src/single_op/memory.cpp [hunk 2]
```diff
@@ -0,0 +1,177 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "shared_test_classes/single_op/memory.hpp"
+
+#include <ie_transformations.hpp>
+#include <transformations/control_flow/unroll_tensor_iterator.hpp>
+#include <functional_test_utils/core_config.hpp>
+#include "ngraph/pass/low_latency.hpp"
```
Comment (vurusovs) [inline src/tests/functional/shared_test_classes/src/single_op/memory.cpp:10]: use `openvino/pass` instead

File: src/tests/functional/shared_test_classes/src/single_op/memory.cpp [hunk 3]
```diff
@@ -0,0 +1,177 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "shared_test_classes/single_op/memory.hpp"
+
+#include <ie_transformations.hpp>
+#include <transformations/control_flow/unroll_tensor_iterator.hpp>
+#include <functional_test_utils/core_config.hpp>
```
Comment (vurusovs) [inline src/tests/functional/shared_test_classes/src/single_op/memory.cpp:9]: looks like not used

File: src/tests/functional/shared_test_classes/src/single_op/memory.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    ov::Shape input_shape;
+    std::string targetDevice;
+    ngraph::helpers::MemoryTransformation transformation;
+    std::tie(transformation, iteration_count, input_shape, model_type, targetDevice) = obj.param;
+
+    std::ostringstream result;
+    result << "transformation=" << transformation << "_";
+    result << "iteration_count=" << iteration_count << "_";
+    result << "IS=" << ov::test::utils::vec2str(input_shape) << "_";
+    result << "netPRC=" << model_type.get_type_name() << "_";
```
Comment (vurusovs) [inline src/tests/functional/shared_test_classes/src/single_op/memory.cpp:38]: netPRC -> modelType

File: src/tests/functional/shared_test_classes/src/single_op/memory.cpp [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    inferRequest = compiledModel.create_infer_request();
+    for (size_t iter = 0; iter <= iteration_count; iter++) {
+        for (const auto& input : inputs) {
+            inferRequest.set_tensor(input.first, input.second);
+        }
+        inferRequest.infer();
+    }
+}
+
+std::vector<ov::Tensor> MemoryLayerTest::calculate_refs() {
```
Comment (vurusovs) [inline src/tests/functional/shared_test_classes/src/single_op/memory.cpp:140]: As I see, the function is a full copy of `SubgraphBaseTest::calculate_refs()`. Please, try to remove it and use basic implementation

**Review by vurusovs — CHANGES_REQUESTED** [review]

File: src/tests/functional/shared_test_classes/src/single_op/memory.cpp (hunk 5 above)
Comment (olpipi) [reply src/tests/functional/shared_test_classes/src/single_op/memory.cpp:140]: Unfortunately not a full copy.
It has additional loop over infer() call
`for (size_t iter = 0; iter <= iteration_count; iter++)`

Comment (vurusovs) [reply src/tests/functional/shared_test_classes/src/single_op/memory.cpp:140]: @olpipi you're right, I though I removed this comment before submitting 😄

File: src/tests/functional/shared_test_classes/src/single_op/memory.cpp (hunk 4 above)
Comment (olpipi) [reply src/tests/functional/shared_test_classes/src/single_op/memory.cpp]: done

File: src/tests/functional/shared_test_classes/src/single_op/memory.cpp (hunk 2 above)
Comment (olpipi) [reply src/tests/functional/shared_test_classes/src/single_op/memory.cpp]: done

File: src/tests/functional/shared_test_classes/include/shared_test_classes/single_op/memory.hpp (hunk 1 above)
Comment (olpipi) [reply src/tests/functional/shared_test_classes/include/shared_test_classes/single_op/memory.hpp]: done

File: src/tests/functional/shared_test_classes/src/single_op/memory.cpp (hunk 3 above)
Comment (olpipi) [reply src/tests/functional/shared_test_classes/src/single_op/memory.cpp]: done

**Comment by olpipi** [comment]
https://openvino-ci.toolbox.iotg.sclab.intel.com/job/private-ci/job/ie/job/build-linux-macos_arm64/5738/

**Review by vurusovs — APPROVED** [review]

**Pull request merged**